src/Storage/Notmuch.hs

```diff
@@ -24,6 +24,27 @@ import Notmuch.Util (bracketT)
 
 import Error
 
+class ManageTags a where
+  setTags :: a -> [T.Text] -> a
+  addTags :: a -> [T.Text] -> a
+  removeTags :: a -> [T.Text] -> a
+  getTags :: a -> [T.Text]
```

Comment:
replace all of this with a lens.  Then if you want the convenience functions define them at the top level.  example:
```haskell
class HasTags a where
  tags :: Lens' a [T.Text]
  writeTags :: (MonadError e m, MonadIO m, AsError e) => FilePath -> a -> m a

setTags, addTags, removeTags :: (HasTags a) => [T.Text] -> a -> a
...
```

Reply:
Great! Why have I not thought of that before. Cheers!